### PR TITLE
Implement Phase 3: Sum Differentiation with Index-Aware Collapse**

### PR DESCRIPTION
**Pull request by jeffreyhorn**
Fixes sum differentiation to correctly collapse when differentiating with respect to concrete indices.

**Changes:**
- Updated `_diff_sum()` to detect and collapse sums when symbolic indices (e.g., "i") match concrete indices (e.g., "i1")
- Added 4 helper functions for sum collapse logic: `_sum_should_collapse()`, `_is_concrete_instance_of()`, `_substitute_sum_indices()`, `_apply_index_substitution()`
- Fixed incorrect behavior: `∂(sum(i, x(i)^2))/∂x(i1)` now returns `2*x(i1)` instead of `Sum(i, 2*x(i))`
- Updated test expectations and added 3 new edge case tests

**Testing:**
- All 20 gradient tests pass ✓
- mypy and ruff checks pass ✓

**Dependencies:**
- Builds on PR #11 (Phase 1) and PR #12 (Phase 2)

### CODE REVIEW

File: src/ad/derivative_rules.py [hunk 1]
```diff
@@ -830,8 +830,160 @@ def _diff_sum(expr: Sum, wrt_var: str, wrt_indices: tuple[str, ...] | None = Non
...
+        return Call(expr.func, new_args)
+    elif isinstance(expr, Sum):
+        # Don't substitute sum's own bound variables, but substitute in body
+        # Filter out sum's bound variables from substitution
+        filtered_sub = {k: v for k, v in substitution.items() if k not in expr.index_sets}
+        new_body = _apply_index_substitution(expr.body, filtered_sub)
+        return Sum(expr.index_sets, new_body)
+    else:
+        # Unknown expression type, return as-is
+        return expr
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/ad/derivative_rules.py]: Add explicit handling for SymbolRef expression
type in _apply_index_substitution function. While SymbolRef represents
scalar symbols without indices (so substitution doesn't change them),
explicitly handling this case improves completeness and consistency
with the dispatcher pattern used in differentiate_expr.

- Added elif isinstance(expr, SymbolRef): return expr case
- Maintains consistency with expression type dispatching pattern
- All tests pass ✓
